### PR TITLE
Generate RPMs and DEBs

### PR DESCRIPTION
**Pull request by Hoverbear**
###  What have you changed?

Add building `rpm`s and `deb`s to the `dist_artifacts` job we're using in our distributed release builds.

As a result of this change we're now using 'prerequisites': https://www.gnu.org/software/make/manual/html_node/Automatic-Prerequisites.html

So you can build any of the distributed packages just by targeting them. Order doesn't matter and it'll do the least work possible.

```bash
make dist/tikv.rpm dist/tikv.deb dist/tikv.tar.gz dist/tikv-docker.tar.gz
```

These packages only install a stub systemd unit and the config template alongside the conventional directory topology.

###  What is the type of the changes?

Engineering/Feature

###  How is the PR tested?
Please select the tests that you ran to verify your changes:
- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

- If there is a document change, please file a PR in ([docs](https://github.com/tikv/website/tree/master/content)) and add the PR number here.
- If this PR should be mentioned in the release note, please update the [release notes](https://github.com/tikv/tikv/blob/master/CHANGELOG.md).

###  Does this PR affect `tidb-ansible`?

If there is a configuration or metrics change, please file a PR in [tidb-ansible](https://github.com/pingcap/tidb-ansible), and add the PR number here.

###  Refer to a related PR or issue link (optional)

#6081 #5777 

###  Any examples? (optional)

```bash
make dist_artifacts
dpkg -i dist/tikv.deb
yum install dist/tikv.rpm
docker load dist/tikv-docker.tar.gz
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -85,6 +85,7 @@ CARGO_TARGET_DIR ?= $(CURDIR)/target
 # Build-time environment, captured for reporting by the application binary
 BUILD_INFO_GIT_FALLBACK := "Unknown (no git or not git repo)"
 BUILD_INFO_RUSTC_FALLBACK := "Unknown"
+export TIKV_BUILD_VERSION = $(shell awk -F ' = ' '$$1 ~ /version/ { gsub(/[\"]/, "", $$2); printf("%s",$$2) }' Cargo.toml)
```
Comment (zhouqiang-cl) [inline Makefile]: why we need this TIKV_BUILD_VERSION

File: etc/deb/preinst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+set -e
+
+if ! getent passwd tikv >/dev/null; then
+        adduser --quiet --system --no-create-home --home /etc/tikv --shell /usr/sbin/nologin tikv
```
Comment (zhouqiang-cl) [inline etc/deb/preinst]: should we put home dir in /etc/tikv?. maybe /home/tikv?

File: etc/rpm/gen-spec.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
+cat <<-EOF
+%global __os_install_post /usr/lib/rpm/brp-compress %{nil}
+%global __spec_install_post %{nil}
+%global debug_package %{nil}
+
+Name:           tikv
+Version:        4.0.0.alpha
```
Comment (zhouqiang-cl) [inline etc/rpm/gen-spec.sh]: can we get it from git command

Comment (Hoverbear) [reply etc/rpm/gen-spec.sh]: Yes! You're right! I need to update this script so we don't need to update this. :)

File: Makefile (hunk 1 above)
Comment (Hoverbear) [reply Makefile]: Seems we don't xD

File: etc/deb/preinst (hunk 2 above)
Comment (Hoverbear) [reply etc/deb/preinst]: I think that's not good packaging practice, and we agreed we'd try our best to follow best practices on TiKV/PD packages. 

Commonly packaged applications will use a location like:

*  `/usr/local/$SERVICE`
* `/var/lib/$SERVICE`

Similarly, this is why the configuration is in `/etc/tikv/config.toml`

Ref:

* PSQL https://www.postgresql.org/docs/9.1/creating-cluster.html
* MySQL https://unix.stackexchange.com/a/113983
* Elasticsearch https://www.elastic.co/guide/en/elasticsearch/reference/2.0/setup-dir-layout.html#default-paths
* Docker: https://docs.docker.com/storage/storagedriver/#the-copy-on-write-cow-strategy
* Mailing list: https://www.redhat.com/archives/rpm-list/2005-November/msg00107.html

File: Makefile (hunk 1 above)
Comment (Hoverbear) [reply Makefile]: Seems on Master the `git describe --tag` gives us 3.0.0-beta which is not right. :(

File: etc/deb/preinst (hunk 2 above)
Comment (tennix) [reply etc/deb/preinst]: Yes, I recommend to use `/var/lib/tikv` as home directory.

File: etc/tikv.service [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+Description=TiKV, a distributed transactional key value store.
+Wants=network-online.target
+
+[Service]
+User=tikv
+ExecStart=/usr/bin/tikv-server --config $CONFIGURATION_DIRECTORY/config.toml
+ExecReload=/bin/kill -HUP $MAINPID
+Restart=on-failure
+KillSignal=SIGINT
+LimitNOFILE=82920
```
Comment (tennix) [inline etc/tikv.service]: By default, nearly all Linux distribution uses 1024 for max number of file descriptors. I'd recommend making the default configuration with `max-open-files` to 1024, so users can start tikv without any problems when they install the package. And we can have a tuning guide to tell user to modify system `ulimit` settings.

File: etc/tikv.service [hunk 5]
```diff
@@ -0,0 +1,16 @@
+[Unit]
```
Comment (tennix) [inline etc/tikv.service]: I think the packaged systemd service file should be placed in `/var/lib/systemd/system`. And if users want to customize the systemd unit file, they should copy the unit file from `/var/lib/systemd/system` to `/etc/systemd/system` or just create /etc/systemd/system/tikv.service.d/overrides.conf. So next time when they upgrade tikv package, the /etc/systemd/system/tikv.service won't screw up.

File: etc/tikv.service (hunk 4 above)
Comment (Hoverbear) [reply etc/tikv.service]: Agree, should we change the default in `/etc/config-template.toml`, or create a config specifically for packages?

File: etc/tikv.service (hunk 5 above)
Comment (Hoverbear) [reply etc/tikv.service]: Agree!

**Comment by Hoverbear** [comment]
Here's a prompt you get on debian if the configs differ:

image

File: etc/tikv.service (hunk 5 above)
Comment (Hoverbear) [reply etc/tikv.service]: This is not correct on Ubuntu/Debian:

```bash
hoverbear@Obsidian:~/git/tikv$ ls /lib/systemd/system/ -l | wc -l
351
hoverbear@Obsidian:~/git/tikv$ ls /var/lib/systemd/system/ -l | wc -l
ls: cannot access '/var/lib/systemd/system/': No such file or directory
```

Comment (Hoverbear) [reply etc/tikv.service]: Also Fedora/Centos:

```bash
hoverbear@Obsidian:~/git/tikv$ docker run --rm centos ls /var/lib/systemd/system -l | wc -l
ls: cannot access '/var/lib/systemd/system': No such file or directory
0
hoverbear@Obsidian:~/git/tikv$ docker run --rm centos ls /lib/systemd/system -l | wc -l
194
```

File: etc/tikv.service [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+Description=TiKV, a distributed transactional key value store.
```
Comment (glorv) [inline etc/tikv.service]: seems this Description discription should be something like `TiKV Service daemon`

**Comment by glorv** [comment]
@Hoverbear  I run similar rpm package generated by tidb-helper. and we I test it in our production environment, service start failed becase the ExecStart commend parameter`$CONFIGURATION_DIRECTORY/config.toml` was parse to empty.

and when I use `systemctl status tidb-server.service -l` to check the status, it report something like:
```
Dec 20 19:23:00 localhost.localdomain systemd[1]: [/usr/lib/systemd/system/tidb-server.service:13] Unknown lvalue 'ConfigurationDirectory' in section 'Service'
```

It seems this kind of config settings donot have good enough compatibility.

File: etc/tikv.service (hunk 4 above)
Comment (sykp241095) [reply etc/tikv.service]: 1024 is the default value

Comment (Hoverbear) [reply etc/tikv.service]: Sadly it isn't:

https://github.com/tikv/tikv/blob/535ac80b66050c3d624838719cc65fe850ee5f9e/etc/config-template.toml#L348
https://github.com/tikv/tikv/blob/535ac80b66050c3d624838719cc65fe850ee5f9e/etc/config-template.toml#L348

File: etc/tikv.service (hunk 6 above)
Comment (Hoverbear) [reply etc/tikv.service]: Agreed!

**Comment by Hoverbear** [comment]
Oh! Wild! I'll fall back to a non-env then. :) Thanks for this great catch.

**Comment by Hoverbear** [comment]
This depends on https://github.com/tikv/tikv/pull/6408, resulting from changes to our component topology.

File: etc/tikv.service (hunk 4 above)
Comment (Hoverbear) [reply etc/tikv.service]: @tennix @sykp241095 do you have further guidance on this?

File: etc/tikv.service (hunk 5 above)
Comment (Hoverbear) [reply etc/tikv.service]: @tennix I am resolving this as your suggestion is unfortunately not correct :(

**Comment by zhouqiang-cl** [comment]
@sykp241095 please test on this pr

**Comment by Hoverbear** [comment]
Closing. I am not longer involved with the project and have no stake in this merging. Feel free to create a new PR.

**Pull request closed without merging**